### PR TITLE
Fix bugs for RTL languages by SCSS in BO

### LINKED ISSUE #17862: .rtlfix files should not be in public folders of BO themes
**Is your feature request related to a problem? Please describe.**
As we decided to avoid pushing the assets on the repo, we need to keep .rtlfix files as they are used to fix some bugs while on RTL languages.

But they should'nt be in the asset's folder, we should extract it somewhere and a backend change has to be done

**Describe the solution you'd like**
Change the path of the rtlfix to another which is not in asset's folder

> Edit by @mparvazi 
## How to get rid of .rtlfix

**1) Delete theme.rtlfix**

**2) Create rtl.scss**
`admin-dev/themes/new-theme/scss/rtl.scss`
The content of rtl.scss is exactly same as theme.rtlfix (it's better to be redesigned for Sass)

**3) Use webpack to build rtl.css**
 `admin-dev/themes/new-theme/.webpack/common.js`

webpack

**3) Add a condition to prevent rtl.css to be transformed by CSSJanus**
`src/Core/Localization/RTL/StylesheetGenerator.php`

proccess

**4) If shop language is RTL then add rtl.css**
`classes/controller/AdminController.php`

shop

### PR DESCRIPTION
**Pull request by mparvazi**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | RTL override rules for BO implemented by SCSS. Vazirmatn font will be used only for Arabic script languages (Arabic/Persian).
| Type?             | improvement
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | no
| Fixed ticket?     | Fixes #17862
| Related PRs       | Nope
| How to test?      | You MUST NOT see any RTL problems in BO (Product V2 is not considered). It's better to be tested on a fresh install.
| Possible impacts? | RTL version of BO

## BC Break
We remove the `.rtlfix` file, it means that if you use it, you'll need to port your changes inside the SCSS stylesheet

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/28277)

### CODE REVIEW

**Review by laurentrousseau918 — APPROVED** [review]

**Comment by NeOMakinG** [comment]
Hey @mparvazi, I appreciate that PR, avoiding the .rtlfix file is really cool, permit us to use SCSS variables and remove some complexity.

I think we should tag this as a BC, I know .rtlfix file shouldn't be modified in production, but I guess some guys may fix things inside it, it's informal, wdyt @PrestaShop/prestashop-maintainers ?

File: admin-dev/themes/default/scss/rtl.scss
Comment (NeOMakinG) [inline admin-dev/themes/default/scss/rtl.scss:209]: All of these font-family props are here because we added some font family to LTR selectors?

Comment (mparvazi) [reply admin-dev/themes/default/scss/rtl.scss:209]: Yes!
And most of them used `!important` in CSS rules.

**Comment by mparvazi** [comment]
> Hey @mparvazi, I appreciate that PR, avoiding the .rtlfix file is really cool, permit us to use SCSS variables and remove some complexity.
> 
> I think we should tag this as a BC, I know .rtlfix file shouldn't be modified in production, but I guess some guys may fix things inside it, it's informal, wdyt @PrestaShop/prestashop-maintainers ?

The final rtl.css content will be as same as theme.rtlfix content.
Some of these rules are no longer needed but I didn't delete these rules.
But I think it's a good idea to to tag this as a BC.

**Review by Progi1984 — APPROVED** [review]

**Comment by matks** [comment]
> > Hey @mparvazi, I appreciate that PR, avoiding the .rtlfix file is really cool, permit us to use SCSS variables and remove some complexity.
> > I think we should tag this as a BC, I know .rtlfix file shouldn't be modified in production, but I guess some guys may fix things inside it, it's informal, wdyt @PrestaShop/prestashop-maintainers ?
> 
> The final rtl.css content will be as same as theme.rtlfix content. Some of these rules are no longer needed but I didn't delete these rules. But I think it's a good idea to to tag this as a BC.

@NeOMakinG is right: some people might have their own theme and have followed the .rtlfix strategy to have a RTL compliant version 😄 . So even if we are able to stop using it, we have to acknowledge people use sometimes their own theme on BO and FO, and this means impactful changes must be announced.

**Review by matks — COMMENTED** [review]
@mparvazi So if I understand it right, you are able to make the BO theme fully RTL compliant with only CSS? No need for custom CSS "just for RTL".

This is a game changer ❤️ !

(I am not able to review your pull request because I'm a very bad frontend developer)

**Review by NeOMakinG — APPROVED** [review]

**Comment by mparvazi** [comment]
Hello @matks 

I checked again!
IMO this PR is not a BC break.

This PR does not stop `.rtlfix` feature in Core, it just stops using of it in BO theme.

**How `.rtlfix` works on Core?**
The `StylesheetGenerator.php` uses this option for BO and FO themes in RTL.
1. In BO when a RTL language added then it transform (with `CSSJanus` plugin) all css files (e.g. theme.css) to RTL version and then will add the content of `.rtlfx` to the man css file (theme_rtl.css).
`theme_rtl.css = transformed(theme.css) + theme.rtlfix`
2. In FO when user want to adopt current theme to RTL (for example classic theme) then StylesheetGenerator.php add the content of `.rtlfix` file (if exists) to transformed css file.
`BO > Theme & Logo > Theme >  Generate RTL stylesheet` (I tested for classic theme and it added `theme.rtlfix` content to `theme_rtl.css`)

**How this PR handles RTL fix rules on BO without `.rtlfix` file?**
All RTL overrides rules are inside `rtl.scss` file. After build the theme with webpack, the `rtl.css` file will be created. 
This file will be include in BO pages after main css file (theme_rtl.css).
```
<link href=".../new-theme/public/theme_rtl.css" rel="stylesheet" type="text/css"/>
<link href=".../new-theme/public/rtl.css" rel="stylesheet" type="text/css"/>
```

**What's the benefit of `rtl.scss` than `theme.rtlfix`?**
1. The `rtl.scss` syntax will be check with Sass parser for errors, the parsed file (rtl.css) is compressed by CSS Minifier, the cross browser rules are added by Sass parser for some rules (-webkit, -moz, ...). 
2. When you want to fix a RTL bug just need to find scss rule in current LTR files and then put it with fix in rtl.scss. Currently you need to find css rule form browser inspector and put it with fix in theme.rtlfix (it's really boring!).
3. You can use `npm run watch` for any changes on-demand for rtl.scss, but currently you need to add override rules in theme.rtlfix and save it, then disable/enable the current RTL language to theme_rtl.css generate with changes (it's really time consuming)

**How theme developers can fix RTL bugs with this new feature?**
Just put rtl.css (RTL override rules) in theme css assets. This file will not transform by CSSJanus because of this change in PR:

**Comment by NeOMakinG** [comment]
I think that it can be considered as a BC, if someone fixed something on the actual content of the`.rtlfix` file, he will definitely need to port his fix inside the current SCSS or remove the SCSS you added or remove unused content from his `rtlfix`

It doesn't hurt anyone to tag it as a BC

**Comment by mparvazi** [comment]
> @mparvazi So if I understand it right, you are able to make the BO theme fully RTL compliant with only CSS? No need for custom CSS "just for RTL".
> 
> This is a game changer ❤️ !
> 
> (I am not able to review your pull request because I'm a very bad frontend developer)

There is two famous plugins for transformation LTR to RTL:
1. [CSSJanus](https://cssjanus.github.io/) (used currently by PS)
2. [RTLCSS](https://rtlcss.com/) (used by Bootstrap 5)

These plugins are really good in transformation but not completely successful on it (as I experienced works fine about 99%)
But some little rules still remains and need to handle them (even Bootstrap RTL has some bugs for RTL languages).
So need to use something like `rtl.scss` to fix these rules.

The PS core uses PHP version of CSSJanus to generate RTL styles.
But hummingird uses webpack version of CSSJanus, so it's RTL adopted (we can name it Auto RTL feature).

In the future, selectors such as [:dir](http://mdn.io/:dir) will help us to fix RTL bugs easily (currently compatible with Firefox only).

**Review by kpodemski — APPROVED** [review]

**Review by sLorenzini — COMMENTED** [review]
hello @mparvazi,

I've tested RTL options with different languages (English / Arabic / Japanese / Russian).
Tested all those languages on pages/elements in Desktop & Mobile screens:

- Dashboard
- Orders > Orders page
- Orders > Orders > Order detail
- Catalog > Products page
- Catalog > Products > Product edit
- Customers > Customer page
- Customers > Customer >  > Customer edit
- Customer Service > Merchandise returns
- Stats 
- Modules > Module manager
- Design > Theme & Logo
- Design > Theme & Logo > Add new theme
- Design > Positions
- Design > Positions > Transplant a module
- Shipping > Carriers
- Shipping > Carriers > Carrier edit
- Payment > Payment Methods 
- Payment > Payment Methods > Configure a payment
- International > Localization
- International > Localization > Languages
- International > Localization > Languages > Edit language
- Shop parameters > Product Settings
- Advanced Parameters > Performance

Clicked on selected list > options are on the right
Tested text field and they are editable 

---

However I've noticed an issue with Arabic language only.

Content in buttons and in tables are not centered vertically.
Font in the main menu are not the same as other languages

English Version with this PR:

english-rtl

Arabic Version with this PR:

arabic-rtl

The English version is fixed in this PR. This language had the same issue as Arabic without this PR.

English Version without this PR:

english-current-develop-version

Could you please check for the Arabic?

Thanks

**Comment by mparvazi** [comment]
Hello @sLorenzini 



I think maybe it's a bug in Vazirmatn font, and it happens when all the characters are not in type of arabic script. 
So let me check with font author.

**Comment by mparvazi** [comment]
@NeOMakinG I need to know your opinion:

The baseline for Latin characters and Arabic characters is same in Vazirmatn:
52527758-383e4180-2cce-11e9-950e-c6fc9bee8aa2
So when all characters in text are Latin and font-face is Vazirmatn problem for line height occurs:
Screenshot from 2022-04-30 16-16-22
We can fix the problem like this:
```
body {
  &.lang-fa,
  &.lang-ar {
    font-family: "Open Sans", Vazirmatn, sans-serif;
  }
}
```
Screenshot from 2022-04-30 16-16-44
So only Arabic charterers will use Vazirmatn and Latin charterers will use "Open Sans" (as like as LTR)
_"Open Sans" used for BO in LTR, this font doesn't support Arabic letters_

Btw font author told me need to work on this problem to find a solution.

**Comment by mparvazi** [comment]
@NeOMakinG

The baseline for Latin characters and Arabic characters is same in Vazirmatn.
This problem usually exists with Persian / Arabic fonts. The reason is that most of the Persian / Arabic letters are below of the baseline. As a result, Persian / Arabic letters often need more space to display than Latin letters.
52527758-383e4180-2cce-11e9-950e-c6fc9bee8aa2
So when all characters in text are Latin and font-face is Vazirmatn, problem for font display occurs:
(Vazirmatn combined with Roboto to display Latin letters)
سفارش‌ها-•-rtl

Screenshot from 2022-04-30 16-16-22

### We can fix the problem by:

1. Using "Open Sans" for Latin characters
```
body {
  &.lang-fa,
  &.lang-ar {
    font-family: "Open Sans", Vazirmatn, sans-serif;
  }
}
```
So only Arabic characters will use Vazirmatn and Latin characters will use "Open Sans" (as like as LTR)
The problem is that "Open Sans" forces rules for line-height, spaces and ... so other issue happens for Arabic letters.

2. The font author told me need to work on this problem to find a solution. It's going to be fixed in next version (I hope!). So we can keep it without any change and wait to be fixed.

IMO option 2 is better!

PS is 100% translated for Persian and 98% for Arabic.
BO is like this for RTL (Orders page):
Screenshot 2022-04-30 at 18-07-07 سفارش‌ها • rtl

**Comment by NeOMakinG** [comment]
Just wait for `Vazirmatn` to be handling the problem.

IMO it's always best to wait for a dependency fix instead of patching it in our solution

@sLorenzini The bug you found is not blocking the PR, it will be naturally fixed in the future by updating `Vazirmatn` when it will get patched, I guess we can merge this :)

**Comment by mparvazi** [comment]
@NeOMakinG 
The problem (Is this a problem? IDTS!) found by @sLorenzini is not related to this PR.

Btw if this is a problem, we already using Vazirmatn in BO with latest version.
So why we should block this PR?

**Comment by NeOMakinG** [comment]
@mparvazi I said that we shouldn't block the PR 😁

**Comment by mparvazi** [comment]
> @mparvazi I said that we shouldn't block the PR grin

Sorry 😅

**Comment by mparvazi** [comment]
Hello @sLorenzini 
For better testing this PR for RTL, please try to test in a fresh install of PS for RTL languages (Arabic / Persian).
This PR doesn't impact LTR version of BO.

**Review by sLorenzini — APPROVED** [review]
hello @mparvazi 

Re-tested RTL options with Persian / Arabic / Latin languages and all good.

I've created also an issue concerning Vazirmatn font to not forget to update this one when it will be possible: https://github.com/PrestaShop/PrestaShop/issues/28452 

QA approved ✅

**Comment by mparvazi** [comment]
Thanks @sLorenzini

**Comment by Progi1984** [comment]
Thanks @mparvazi & @sLorenzini

**Pull request merged**
